File: backend/src/app/securepoint/ssh.py
"""SSH transport for Securepoint: run ``swanctl --raw`` on the box and parse it.

The dashboard authenticates with a single ed25519 key (``DASH_SSH_PRIVATE_KEY``);
its public half is installed on each box (see docs/securepoint-ssh.md). This is
what gives the pull path the IKE cookies + ESP SPIs + byte counters that the
spcgi API never exposes — i.e. the data needed to pair tunnel ends across NAT.

Host-key handling is trust-on-first-use, fail-closed: command-running connections
refuse to proceed unless a pinned ``host_key`` is present and matches the server;
``probe_host_key`` is the only path that connects unpinned, to capture it for
storage. Until a key is pinned, callers fall back to the spcgi API.
"""

from __future__ import annotations

import asyncio
import re
from dataclasses import dataclass

import asyncssh

from app.securepoint.swanctl import ipsec_status_from_swanctl
from app.xsense.schemas import DiagnosisSection, IPsecServiceStatus

# Diagnose runs several commands + a ping + a syslog dump; it's a user-triggered
# one-off (no wait_for around it), so a generous timeout is fine.
_DIAG_TIMEOUT = 25.0
_SEC = "@@SEC@@"
# Connection names are safe shell tokens; reject anything else before interpolating.
_SAFE_NAME = re.compile(r"[A-Za-z0-9._:-]{1,128}")

# Budget: the global VPN overview wraps ipsec_status() in a wait_for(_FETCH_TIMEOUT)
# (views/routes.py, 8s) and silently drops the instance on timeout. Keep the whole
# SSH round-trip under that: connect + the two swanctl runs (executed concurrently).
_CONNECT_TIMEOUT = 4.0
_CMD_TIMEOUT = 3.0


class SecurepointSSHError(RuntimeError):
    pass


@dataclass(frozen=True)
class SSHConfig:
    """Per-instance SSH access for swanctl enrichment (private key already decrypted)."""

    host: str
    port: int
    user: str
    private_key: str
    host_key: str | None = None


def _key_blob(openssh_line: str) -> str:
    """The base64 blob of an ``ssh-ed25519 AAAA... [comment]`` line (identity part)."""
    parts = openssh_line.split()
    return parts[1] if len(parts) >= 2 else openssh_line.strip()


async def _connect(
    host: str,
    port: int,
    user: str,
    private_key_pem: str,
    host_key: str | None,
    *,
    require_host_key: bool = True,
) -> tuple[asyncssh.SSHClientConnection, str]:
    """Open a connection; return (conn, server_host_key_openssh).

    Fail-closed on the host key: a command-running connection (``require_host_key``,
    the default) refuses to proceed unless a pinned ``host_key`` is supplied and
    matches the server. Only ``probe_host_key`` connects unpinned (TOFU capture), by
    passing ``require_host_key=False``. Without this, ``known_hosts=None`` would
    accept any host key, letting an on-path attacker impersonate the box and feed
    fabricated swanctl/diagnostic output.
    """
    if require_host_key and not host_key:
        raise SecurepointSSHError(
            "SSH host key not pinned — refusing to connect unverified "
            "(enrichment falls back to the spcgi API until the key is captured)"
        )
    if not private_key_pem.strip():
        raise SecurepointSSHError("no SSH private key configured (DASH_SSH_PRIVATE_KEY)")
    try:
        client_key = asyncssh.import_private_key(private_key_pem)
    except (asyncssh.KeyImportError, ValueError) as exc:
        raise SecurepointSSHError(f"bad SSH private key: {exc}") from exc

    try:
        # known_hosts=None: skip asyncssh's lookup; we verify the host key ourselves
        # right after the handshake and before running any command.
        conn = await asyncio.wait_for(
            asyncssh.connect(
                host,
                port=port,
                username=user,
                client_keys=[client_key],
                known_hosts=None,
            ),
            timeout=_CONNECT_TIMEOUT,
        )
    except (TimeoutError, OSError, asyncssh.Error) as exc:
        raise SecurepointSSHError(f"SSH connect {user}@{host}:{port} failed: {exc}") from exc

    server_key = conn.get_server_host_key()
    server_line = server_key.export_public_key().decode().strip() if server_key else ""
    if host_key and _key_blob(server_line) != _key_blob(host_key):
        conn.close()
        await conn.wait_closed()
        raise SecurepointSSHError("SSH host key mismatch (possible MITM) — refused")
    return conn, server_line


async def probe_host_key(host: str, port: int, user: str, private_key_pem: str) -> str:
    """Connect once (unpinned) and return the box's host key for storage (TOFU)."""
    conn, server_line = await _connect(
        host, port, user, private_key_pem, host_key=None, require_host_key=False
    )
    conn.close()
    await conn.wait_closed()
    return server_line


async def fetch_ipsec_status(
    host: str,
    port: int,
    user: str,
    private_key_pem: str,
    host_key: str | None,
    *,
    running: bool,
) -> IPsecServiceStatus:
    """Run swanctl over SSH and parse it into the shared IPsec DTO (with SPIs)."""
    conn, _ = await _connect(host, port, user, private_key_pem, host_key)
    try:
        sas = await asyncio.wait_for(conn.run("swanctl --list-sas --raw"), timeout=_CMD_TIMEOUT)
        conns = await asyncio.wait_for(conn.run("swanctl --list-conns --raw"), timeout=_CMD_TIMEOUT)
    except (TimeoutError, asyncssh.Error) as exc:
        raise SecurepointSSHError(f"swanctl over SSH failed: {exc}") from exc
    finally:
        conn.close()
        await conn.wait_closed()
    return ipsec_status_from_swanctl(
        str(sas.stdout or ""), str(conns.stdout or ""), running=running
    )


def _diag_script(name: str) -> str:
    """Shell run over one SSH session that emits ``@@SEC@@<title>``-delimited blocks.

    Gathers as much as the box exposes: the connection config, live SAs, installed
    policies, the recent charon log (vici-poll noise stripped — failure lines like
    NO_PROPOSAL_CHOSEN / AUTHENTICATION_FAILED / 'giving up' live here regardless of
    whether they carry the conn name), and a one-shot peer-reachability ping.
    """
    return (
        f"N='{name}'\n"
        f"echo '{_SEC}Connection config (swanctl --list-conns)'\n"
        "swanctl --list-conns 2>&1\n"
        f"echo '{_SEC}Live IKE / CHILD SAs (swanctl --list-sas)'\n"
        'swanctl --list-sas --ike "$N" 2>&1\n'
        f"echo '{_SEC}Recent IPsec log (charon)'\n"
        # syslog is a padded `msgid|date|app|pid|msg` table — match the app field
        # (col 3) robustly, drop our own vici-poll noise, cap the tail.
        "(echo 'syslog get' | spcli 2>/dev/null) "
        "| awk -F'|' 'NR>2 && $3 ~ /charon/ && $0 !~ /\\[CFG\\] vici client/' "
        "| tail -n 300\n"
        f"echo '{_SEC}Peer reachability'\n"
        "REMOTE=$(swanctl --list-conns --raw 2>/dev/null | grep -oE 'remote_addrs=\\[[^]]*' "
        "| grep -oE '[0-9]+\\.[0-9]+\\.[0-9]+\\.[0-9]+' | head -n1)\n"
        '[ -z "$REMOTE" ] && REMOTE=$(swanctl --list-sas --ike "$N" --raw 2>/dev/null '
        "| grep -oE 'remote-host=[0-9.]+' | head -n1 | cut -d= -f2)\n"
        '[ -n "$REMOTE" ] && { echo "ping $REMOTE:"; ping -c 2 -w 4 "$REMOTE" 2>&1; } '
        '|| echo "no concrete peer IP (remote=%any / responder-only) — nothing to ping"\n'
    )


def _parse_sections(out: str) -> list[DiagnosisSection]:
    sections: list[DiagnosisSection] = []
    title: str | None = None
    buf: list[str] = []
    for line in out.splitlines():
        if line.startswith(_SEC):
            if title is not None:
                sections.append(DiagnosisSection(title=title, content="\n".join(buf).strip()))
            title = line[len(_SEC) :].strip()
            buf = []
        elif title is not None:
            buf.append(line)
    if title is not None:
        sections.append(DiagnosisSection(title=title, content="\n".join(buf).strip()))
    return sections


async def fetch_diagnosis(
    host: str,
    port: int,
    user: str,
    private_key_pem: str,
    host_key: str | None,
    tunnel_id: str,
) -> list[DiagnosisSection]:
    """Gather a readable per-tunnel diagnostic bundle over one SSH session."""
    if not _SAFE_NAME.fullmatch(tunnel_id):
        raise SecurepointSSHError(f"unsafe tunnel id: {tunnel_id!r}")
    conn, _ = await _connect(host, port, user, private_key_pem, host_key)
    try:
        res = await asyncio.wait_for(conn.run(_diag_script(tunnel_id)), timeout=_DIAG_TIMEOUT)
    except (TimeoutError, asyncssh.Error) as exc:
        raise SecurepointSSHError(f"diagnose over SSH failed: {exc}") from exc
    finally:
        conn.close()
        await conn.wait_closed()
    return _parse_sections(str(res.stdout or ""))

File: orbit/lib/orbit/securepoint/ssh.ex
defmodule Orbit.Securepoint.SSH do
  @moduledoc """
  SSH transport for Securepoint: run `swanctl --raw` on the box and parse it.

  Port of the deleted `backend/src/app/securepoint/ssh.py`. The dashboard
  authenticates with one ed25519 key per instance (`just gen-ssh-key`, private
  half stored Fernet-encrypted in `instances.ssh_key_enc`, public half installed
  on the box — see `docs/securepoint-ssh.md`). This is what gives the pull path
  the IKE cookies, ESP SPIs and byte counters that the spcgi API never exposes,
  i.e. the data needed to pair tunnel ends across NAT.

  ## Host-key handling is trust-on-first-use, FAIL-CLOSED

  A command-running connection refuses to proceed unless a pinned host key is
  present AND matches the server. `probe_host_key/1` is the only path that
  connects unpinned, and only to capture the key for storage. Until a key is
  pinned, callers fall back to the spcgi API.

  Do not "simplify" this into accepting any host key: without verification an
  on-path attacker impersonates the box and feeds fabricated swanctl output,
  which the dashboard would render as tunnel state and alert on.

  Erlang's `:ssh` (no external dependency) does the work. It must stay listed in
  `extra_applications` in mix.exs — Elixir prunes the code path to the declared
  applications, so without it `:ssh` is missing from `mix run` AND the release.
  """

  alias Orbit.Securepoint.Swanctl

  @connect_timeout 8_000
  @cmd_timeout 5_000

  defmodule Config do
    @moduledoc "Per-instance SSH access (private key already decrypted)."
    @enforce_keys [:host, :port, :user, :private_key]
    defstruct [:host, :port, :user, :private_key, host_key: nil]
  end

  defmodule KeyCb do
    @moduledoc """
    `:ssh_client_key_api` callback: supplies the client key and decides the host
    key. Erlang has no "give me the server key after the handshake" hook, so the
    comparison happens here — which is also the only place it CAN be
    fail-closed, because returning false aborts before any channel opens.
    """
    @behaviour :ssh_client_key_api

    @impl true
    def user_key(_algo, opts) do
      {:ok, Keyword.fetch!(opts[:key_cb_private], :priv_key)}
    end

    @impl true
    def is_host_key(key, _host, _port, _algo, opts) do
      priv = opts[:key_cb_private]

      case Keyword.get(priv, :pinned) do
        nil ->
          # TOFU capture only — reached via probe_host_key/1, never while
          # running a command.
          send(
            Keyword.fetch!(priv, :reply_to),
            {:host_key, Orbit.Securepoint.SSH.encode_key(key)}
          )

          true

        pinned ->
          Orbit.Securepoint.SSH.same_key?(Orbit.Securepoint.SSH.encode_key(key), pinned)
      end
    end

    @impl true
    def add_host_key(_host, _port, _key, _opts), do: :ok
  end

  # -- public ---------------------------------------------------------------

  @doc """
  Run swanctl over SSH and return the `ipsec` hub section (with SPIs).

  `running` is the service state the caller already knows from the spcgi API —
  swanctl answering at all implies the daemon is up, but the caller stays the
  authority.
  """
  @spec fetch_ipsec_status(%Config{}, boolean()) :: {:ok, map()} | {:error, String.t()}
  def fetch_ipsec_status(%Config{} = cfg, running) when is_boolean(running) do
    with {:ok, conn} <- connect(cfg) do
      try do
        with {:ok, sas} <- exec(conn, ~c"swanctl --list-sas --raw"),
             {:ok, conns} <- exec(conn, ~c"swanctl --list-conns --raw") do
          {:ok, Swanctl.status(sas, conns, running)}
        end
      after
        :ssh.close(conn)
      end
    end
  end

  @doc """
  Connect once WITHOUT a pinned key and return the box's host key for storage.

  The only unpinned path in this module (trust on first use). Everything else
  refuses to run unverified.
  """
  @spec probe_host_key(%Config{}) :: {:ok, String.t()} | {:error, String.t()}
  def probe_host_key(%Config{} = cfg) do
    with {:ok, conn} <- connect(%{cfg | host_key: nil}, require_host_key: false) do
      :ssh.close(conn)

      receive do
        {:host_key, line} -> {:ok, line}
      after
        0 -> {:error, "connected but the server host key was never offered"}
      end
    end
  end

  @doc false
  def connect(%Config{} = cfg, opts \\ []) do
    if Keyword.get(opts, :require_host_key, true) and blank?(cfg.host_key) do
      {:error,
       "SSH host key not pinned — refusing to connect unverified " <>
         "(enrichment falls back to the spcgi API until the key is captured)"}
    else
      do_connect(cfg)
    end
  end

  # The stored line names its algorithm; PREFER that family so the server
  # presents the key we can actually verify. A box commonly offers several host
  # keys and Erlang would otherwise negotiate by its own order — pinned RSA vs
  # negotiated ECDSA is a guaranteed mismatch and a silent fail-closed refusal.
  #
  # It must be `modify_algorithms [prepend:]`, and both alternatives were
  # measured against a live box:
  #   pref_public_key_algs            -> the USER-auth key list. Leaves the host
  #                                      key untouched AND breaks publickey auth
  #                                      by excluding our ed25519 client key.
  #   preferred_algorithms public_key -> REPLACES the list. Gets the right host
  #                                      key, but drops ed25519 with it, so auth
  #                                      fails.
  #   modify_algorithms prepend       -> reorders only. Right host key, auth
  #                                      still works. This one.
  # RSA additionally accepts the SHA-2 signature algorithms: same key material,
  # newer signing.
  defp host_key_algs(nil), do: []

  defp host_key_algs(line) when is_binary(line) do
    case line |> String.trim() |> String.split(~r/\s+/) |> List.first() do
      "ssh-rsa" ->
        [
          modify_algorithms: [
            prepend: [public_key: [:"rsa-sha2-512", :"rsa-sha2-256", :"ssh-rsa"]]
          ]
        ]

      "ssh-ed25519" ->
        [modify_algorithms: [prepend: [public_key: [:"ssh-ed25519"]]]]

      "ecdsa-sha2-nistp256" ->
        [modify_algorithms: [prepend: [public_key: [:"ecdsa-sha2-nistp256"]]]]

      "ecdsa-sha2-nistp384" ->
        [modify_algorithms: [prepend: [public_key: [:"ecdsa-sha2-nistp384"]]]]

      "ecdsa-sha2-nistp521" ->
        [modify_algorithms: [prepend: [public_key: [:"ecdsa-sha2-nistp521"]]]]

      _ ->
        []
    end
  end

  defp host_key_algs(_), do: []

  defp blank?(nil), do: true
  defp blank?(""), do: true
  defp blank?(v) when is_binary(v), do: String.trim(v) == ""
  defp blank?(_), do: false

  defp do_connect(%Config{} = cfg) do
    with {:ok, priv_key} <- decode_private_key(cfg.private_key) do
      key_cb_private =
        [priv_key: priv_key, reply_to: self()] ++
          if(cfg.host_key, do: [pinned: cfg.host_key], else: [])

      opts =
        [
          user: String.to_charlist(cfg.user || "root"),
          auth_methods: ~c"publickey",
          key_cb: {KeyCb, key_cb_private}
          # Pin the HOST-KEY ALGORITHM to the one we stored, or the comparison is
          # a coin toss: a box commonly offers several host keys (RSA + ECDSA +
          # ed25519) and Erlang would negotiate by its own preference. Pinning an
          # RSA key while Erlang picks ECDSA yields a guaranteed blob mismatch and
          # a fail-closed refusal — the enrichment then never runs, silently.
          # Observed on a live box: pinned ssh-rsa, negotiated ecdsa-sha2-nistp256.
        ] ++
          host_key_algs(cfg.host_key) ++
          [
            silently_accept_hosts: false,
            user_interaction: false,
            save_accepted_host: false,
            quiet_mode: true,
            connect_timeout: @connect_timeout
          ]

      case :ssh.connect(String.to_charlist(cfg.host), cfg.port || 22, opts, @connect_timeout) do
        {:ok, conn} ->
          {:ok, conn}

        {:error, reason} ->
          {:error, "SSH connect #{cfg.user}@#{cfg.host}:#{cfg.port} failed: #{describe(reason)}"}
      end
    end
  end

  @doc false
  def exec(conn, command) do
    case :ssh_connection.session_channel(conn, @cmd_timeout) do
      {:ok, chan} ->
        :ssh_connection.exec(conn, chan, command, @cmd_timeout)
        collect(conn, chan, [])

      {:error, reason} ->
        {:error, "swanctl over SSH failed: #{describe(reason)}"}
    end
  end

  defp collect(conn, chan, acc) do
    receive do
      # type 0 = stdout; 1 = stderr, which swanctl uses for warnings we ignore.
      {:ssh_cm, ^conn, {:data, ^chan, 0, data}} -> collect(conn, chan, [data | acc])
      {:ssh_cm, ^conn, {:data, ^chan, _type, _data}} -> collect(conn, chan, acc)
      {:ssh_cm, ^conn, {:eof, ^chan}} -> collect(conn, chan, acc)
      {:ssh_cm, ^conn, {:exit_status, ^chan, _status}} -> collect(conn, chan, acc)
      {:ssh_cm, ^conn, {:closed, ^chan}} -> {:ok, acc |> Enum.reverse() |> IO.iodata_to_binary()}
      {:ssh_cm, ^conn, _other} -> collect(conn, chan, acc)
    after
      @cmd_timeout ->
        :ssh_connection.close(conn, chan)
        {:error, "swanctl over SSH timed out after #{@cmd_timeout}ms"}
    end
  end

  @doc """
  The base64 blob of an `ssh-ed25519 AAAA… [comment]` line.

  Comparison is on the identity part only: the comment is free text the box may
  rewrite, and the algorithm prefix is implied by the blob.
  """
  def key_blob(line) when is_binary(line) do
    case String.split(String.trim(line), ~r/\s+/) do
      [_algo, blob | _] -> blob
      _ -> String.trim(line)
    end
  end

  @doc false
  def same_key?(a, b), do: key_blob(a) == key_blob(b)

  @doc false
  def encode_key(key) do
    [{key, []}] |> :ssh_file.encode(:openssh_key) |> to_string() |> String.trim()
  rescue
    _ -> ""
  end

  @doc false
  def decode_private_key(pem) when is_binary(pem) do
    if String.trim(pem) == "" do
      {:error, "no SSH private key configured for this instance"}
    else
      case :ssh_file.decode(pem, :openssh_key_v1) do
        [{key, _attrs} | _] -> {:ok, key}
        _ -> {:error, "bad SSH private key: could not decode"}
      end
    end
  rescue
    e -> {:error, "bad SSH private key: #{Exception.message(e)}"}
  end

  def decode_private_key(_), do: {:error, "no SSH private key configured for this instance"}

  defp describe(reason) when is_binary(reason), do: reason
  defp describe(reason) when is_list(reason), do: to_string(reason)
  defp describe(reason), do: inspect(reason)
end
